File: apps/mobile/lib/core/utils/location_fuzzing.dart
import 'dart:math';

// Randomize lat/lng within ±3km radius using uniform disk sampling.
// 1° lat ≈ 111km | 1° lng ≈ 111km * cos(lat_rad)
(double lat, double lng) fuzzLocation(double lat, double lng) {
  final rng = Random.secure();
  const radiusKm = 3.0;
  final u = rng.nextDouble();
  final v = rng.nextDouble();
  final w = radiusKm / 111.0 * sqrt(u);
  final t = 2 * pi * v;
  return (lat + w * cos(t), lng + w * sin(t) / cos(lat * pi / 180));
}

File: functions/src/locationFuzzing.ts
import { onCall, HttpsError } from 'firebase-functions/v2/https';
import { logger } from 'firebase-functions/v2';

interface FuzzRequest {
  lat: number;
  lng: number;
}

interface FuzzResponse {
  lat: number;
  lng: number;
}

// Applies ±3km radius randomization to submitted coordinates.
// Called by the client before writing a report to Firestore.
export const fuzzReportLocation = onCall<FuzzRequest, Promise<FuzzResponse>>({ region: 'asia-southeast1' }, async (request) => {
  if (!request.auth) {
    throw new HttpsError('unauthenticated', 'Must be signed in to submit a report.');
  }

  const { lat, lng } = request.data;
  if (typeof lat !== 'number' || typeof lng !== 'number') {
    throw new HttpsError('invalid-argument', 'lat and lng must be numbers.');
  }

  const radiusKm = 3.0;
  const u = Math.random();
  const v = Math.random();
  const w = (radiusKm / 111.0) * Math.sqrt(u);
  const t = 2 * Math.PI * v;

  const fuzzedLat = lat + w * Math.cos(t);
  const fuzzedLng = lng + (w * Math.sin(t)) / Math.cos((lat * Math.PI) / 180);

  logger.info('fuzzReportLocation: applied ±3km fuzz', { originalLat: lat, originalLng: lng });

  return { lat: fuzzedLat, lng: fuzzedLng };
});
